cmd/registry/controller/controller.go

```diff
@@ -144,107 +136,249 @@ func generateActions(
+	for _, targetResource := range resourceList {
+		visited[targetResource.GetResourceName().GetParent()] = true
 
-		takeAction := false
+		takeAction, err := needsUpdate(
+			targetResource.GetResourceName(),
+			targetResource.GetUpdateTimestamp(),
+			dependencyMaps,
+			generatedResource,
+			false,
+		)
 
-		// Evaluate this resource against each dependency source pattern
-		for i, dependency := range generatedResource.Dependencies {
-			dMap := dependencyMaps[i]
-			// Get the group to look for in dependencyMap
-			group, err := getGroupKey(dependency.Pattern, resource)
+		if err != nil {
+			log.Errorf(ctx, "%s", err)
+			continue
+		}
+
+		if takeAction {
+			cmd, err := generateCommand(generatedResource.Action, targetResource.GetResourceName().String())
 			if err != nil {
-				return nil, fmt.Errorf("cannot match resource with dependency. Error: %s", err.Error())
+				return nil, nil, fmt.Errorf("Cannot generate command: %s", err)
 			}
-
-			if maxUpdateTime, ok := dMap[group]; ok {
-				// Take action if dependency timestamp is later than resource timestamp
-				if maxUpdateTime.After(resourceTime) {
-					takeAction = true
-				}
-				visited[group] = true
-			} else {
-				// For a given resource, each of it's defined dependency group should be present.
-				// If any one of the dependency groups is missing, avoid calculating any action for the resource
-				takeAction = false
-				break
+			a := &Action{
+				Command:           cmd,
+				GeneratedResource: targetResource.GetResourceName().String(),
+				RequiresReceipt:   generatedResource.Receipt,
 			}
+			actions = append(actions, a)
+		}
+
+	}
+
+	return actions, visited, nil
+}
+
+// For the target resources which do not exist in the registry yet,
+//we will use the parent resources to derive which new target resources should be created.
+func generateCreateActions(
+	ctx context.Context,
+	client connection.Client,
+	resourcePattern string,
+	dependencyMaps []map[string]time.Time,
+	generatedResource *rpc.GeneratedResource,
+	visited map[string]bool) ([]*Action, error) {
+
+	actions := make([]*Action, 0)
+
+	parsedResourcePattern, err := parseResourcePattern(resourcePattern)
+	if err != nil {
+		return nil, err
+	}
+
+	parentName := parsedResourcePattern.GetParent()
+
+	// If parent is a project, we can't list projects since this is registry client command.
+	// Since the manifest definition is scoped  only for a particular project,
+	// there will be only one target resource in this case.
+	// There are two cases where this might happen:
+	// 1. Target resource is a project level artifact "projects/demo/locations/global/artifacts/serach-index"
+	//    extracted parent will be "projects/demo/locations/global"
+	// 1. Target resource is an api "projects/demo/locations/global/apis/petstore"
+	//    extracted parent will be "projects/demo/locations/global"
+	if strings.HasSuffix(parentName, "locations/global") {
```

Comment:
One more future optimization in my mind is to converge the `updateActions` and `createActions` methods, where we fetch the parent only if the target resource is absent and not go through all the parents.

Reply:
Suppose this is listing a few thousand specs. Will there be a perceptible difference in performance by looping over it again?

There are other much slower operations (e.g. unnecessary server requests) that we can avoid but choose not to because it's fast enough as-is. Here it matters much less, so I think we should favor code clarity and maintainability.

Reply:
Yes, looping over twice is a concern but I don't have any proof regarding the performance impact. I am completely in favor of code clarity and maintainability, I am just saying that I would like to refactor this in a separate effort where I can look for options to converge both the `parent` data types and also converge the `create` and `update` functions. Hence, the tracking issue: https://github.com/apigee/registry/issues/485

Unrelated, but for multiple server requests, the optimization plan is to introduce caching. Hence, the code right now is doing multiple calls to the registry with exactly the same query, but it should be solved with caching.